packages/driver/src/cy/assertions.js

```diff
@@ -0,0 +1,509 @@
+      let cmd = cmds[i]
+      const setCommandLog = (log) => {
+        // our next log may not be an assertion
+        // due to page events so make sure we wait
+        // until we see page events
+        if (log.get('name') !== 'assert') {
+          return
+        }
+
+        // when we do immediately unbind this function
+        state('onBeforeLog', null)
+
+        const insertNewLog = (log) => {
+          cmd.log(log)
+
+          return options.assertions.push(log)
+        }
+
+        // its possible a single 'should' will assert multiple
+        // things such as the case with have.property. we can
+        // detect when that is happening because cmd will be null.
+        // if thats the case then just set cmd to be the previous
+        // cmd and do not increase 'i'
+        // this will prevent 2 logs from ever showing up but still
+        // provide errors when the 1st assertion fails.
+        if (!cmd) {
+          cmd = cmds[i - 1]
+        } else {
+          i += 1
+        }
+
+        // if our command has a function argument
+        // then we know it may contain multiple
+        // assertions
+        if (cmdHasFunctionArg(cmd)) {
+          let index = cmd.get('assertionIndex')
+          const assertions = cmd.get('assertions')
+
+          // always increase the assertionIndex
+          // so our next assertion matches up
+          // to the correct index
+          const incrementIndex = () => {
+            return cmd.set('assertionIndex', index += 1)
+          }
+
+          // if we dont have an assertion at this
+          // index then insert a new log
+          const assertion = assertions[index]
+
+          if (!assertion) {
+            assertions.push(log)
+            incrementIndex()
+
+            return insertNewLog(log)
+          }
+
+          // else just merge this log
+          // into the previous assertion log
+          incrementIndex()
+          assertion.merge(log)
+
+          // dont output a new log
+          return false
+        }
+
+        // if we already have a log
+        // then just update its attrs from
+        // the new log
+        const l = cmd.getLastLog()
+
+        if (l) {
+          l.merge(log)
+
+          // and make sure we return false
+          // which prevents a new log from
+          // being added
+          return false
+        }
+
+        return insertNewLog(log)
+      }
+
+      state('onBeforeLog', setCommandLog)
+
+      // send verify=true as the last arg
+      return assertFn.apply(this, args.concat(true))
+    }
+
+    const fns = injectAssertionFns(cmds)
+
+    const subjects = []
+
+    // iterate through each subject
+    // and force the assertion to return
+    // this value so it does not get
+    // invoked again
+    const setSubjectAndSkip = () => {
+      subjects.forEach((subject, i) => {
+        const cmd = cmds[i]
+
+        cmd.set('subject', subject)
+        cmd.skip()
+      })
+    }
```

Comment:
Can we define an explicit `return` statement here to avoid confusion.

Reply:
This function is a translation of this:

```coffee
## iterate through each subject
## and force the assertion to return
## this value so it does not get
## invoked again
setSubjectAndSkip = ->
  for subject, i in subjects
    cmd  = cmds[i]
    cmd.set("subject", subject)
    cmd.skip()
```

```js
//# iterate through each subject
//# and force the assertion to return
//# this value so it does not get
//# invoked again
const setSubjectAndSkip = () => (() => {
  const result = [];
  for (i = 0; i < subjects.length; i++) {
    subject = subjects[i];
    const cmd  = cmds[i];
    cmd.set("subject", subject);
    result.push(cmd.skip());
  }
  return result;
})();
```

As I checked the usage of this function, it is only used once in around the line 384:

```js
    .then(() => {
      restore()

      setSubjectAndSkip()

      finishAssertions(options.assertions)

      return onPassFn()
    })
```

It doesn't return anything. And `cmd.skip()` just sets the `skip` property and returns the Command object. 

I concluded that it is a setter function. So, it doesn't return anything. 

That's why I used `forEach` that returns `void`:

```ts
// forEach type
forEach(callbackfn: (value: any, index: number, array: any[]) => void, thisArg?: any): void
```

I guess it will make things more confusing by adding meaningless `return` here.

Reply:
`cmd.skip()` returns `Command` object. So, I make it return the `cmds` like below:

```js
    const setSubjectAndSkip = () => {
      subjects.forEach((subject, i) => {
        const cmd = cmds[i]

        cmd.set('subject', subject)
        cmd.skip()
      })

      return cmds
    }
```